sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java

```diff
@@ -107,17 +104,19 @@
     BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
     Table targetTable = getTargetTable(bqOptions);
 
-    String tableReferenceId = "";
+    ReadSession.Builder readSessionBuilder = ReadSession.newBuilder();
     if (targetTable != null) {
-      tableReferenceId = BigQueryHelpers.toTableResourceName(targetTable.getTableReference());
+      readSessionBuilder.setTable(
+          BigQueryHelpers.toTableResourceName(targetTable.getTableReference()));
     } else {
       // If the table does not exist targetTable will be null.
       // Construct the table id if we can generate it. For error recording/logging.
-      tableReferenceId = getTargetTableId(bqOptions);
+      @Nullable String tableReferenceId = getTargetTableId(bqOptions);
```

Comment:
I don't understand why we would ever want to return null from a call to `getTargetTableId`. Either the table should exist and be readable or we should have raised an exception prior to this point.

Reply:
Yea I found the design less than idea. There are two subclasses of this and one of them implements the method by just returning `null`. Basically it is a difference between a read of a table vs a read of a query.

Reply:
(so I've discovered via users that a query-based BigQueryIO read always NPEs on line 119 from before the diff so this is pretty severe)